Compatibility rework 2

rankynbass opened the pull request:
Replacement compatibility rework to go with PR https://github.com/goatcorp/FFXIVQuickLauncher/pull/1366.

This allows for the selecting of different wine versions, different dxvk versions, and hud options to all be managed on the XL.Core side. The values are then passed to CompatibilityTools. This primarily allows for updating wine or dxvk without touching XL.Common.Unix.

Replaces https://github.com/goatcorp/XIVLauncher.Core/pull/59

Here's a flatpak of the launcher, for testing: https://github.com/rankynbass/XIVLauncher.Core/releases/tag/compatrework-v2

rankynbass commented:
Wine folders can now be dropped in `compatibilitytools/wine` and dxvk folders can be dropped in `compatibilitytool/dxvk`, and they will now appear in the drop down lists. This basically replaces an enum with a dictionary, which is a log more flexible and requires a lot fewer switch statements. Should be easier to update when new versions of wine or dxvk are released.
![xlcore-custom-wine-versions-2](https://github.com/goatcorp/XIVLauncher.Core/assets/109918887/37c0d55d-3a41-41ef-b9c4-44e4335a622e)

yaomtc commented:
[Wine 9 was just released](https://gitlab.winehq.org/wine/wine/-/releases/wine-9.0) with an experimental Wayland driver.  Until such a point as the Wayland driver becomes enabled by default, it might be a good checkbox to add? I haven't tested it yet (waiting for Arch package) but if enabled in the registry it might solve some issues [such as mine](https://bugs.kde.org/show_bug.cgi?id=458233). Will let you know how that goes.

rankynbass commented:
Updated for the XL.Common.Unix merge.

marzent reviewed (changes requested):
Conceptually I find it makes more sense to have `OSInfo` live in `XIVLauncher.Common` (this also gets rid of passing around that extra Flatpak bool).

Given that XL changes are rather hard though, it might be worth to put it for now into something like `XIVLauncher.Common2`, which would also allow to extend the existing interface more easily in the future (including auto-launch features)

rankynbass commented:
Honestly, OSInfo could just be renamed DistroInfo, because that's all it's really doing. Yeah, it has checks for other OSes, but those aren't actually used for anything. It could just as easily be an IsLinux bool instead of an a platform enum. As for the flatpak bool, tbh it's not used for anything either. I mainly added it just in case, but never found any reason to use it in CompatibilityTools. All of the flatpak stuff has to be worked out before it gets passed to CompatibilityTools anyway.

marzent commented:
If the Flatpak bool is truly not needed, then it should probably just be removed honestly.

But even as DistroInfo it does make much more sense to have it conceptually there (and could for example greatly simplify `MangoHudIsInstalled`).

In this MR as it is right now, it is used however much more like a more advanced `OSVersion`.

rankynbass commented:
I've moved OSInfo into XL.Common.Unix and renamed it LinuxInfo. I've also added library paths to LinuxInfo, so that there's a List<string> of paths to search for common library files. I've adjusted the MangoHudIsInstalled function to make use of it, and plan to use it for finding DLSS libraries in the future, when that gets added.

I've also added Snap handling, since it's possible XLCore will be run inside of Snap Steam using XLM on Ubuntu distros.

Blooym commented:
Replaced by #221 (?)

rankynbass commented:
Closing in favor of #221.
